packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts

```diff
@@ -36,6 +44,91 @@ export async function migrateCredentialExchangeRecordToV0_4<Agent extends BaseAg
   }
 }
 
+const holderCredentialStates = [
+  CredentialState.Declined,
+  CredentialState.ProposalSent,
+  CredentialState.OfferReceived,
+  CredentialState.RequestSent,
+  CredentialState.CredentialReceived,
+]
+
+export async function getCredentialRole(agent: BaseAgent, credentialRecord: CredentialExchangeRecord) {
+  // Credentials will only have a value when a credential is received, meaning we're the holder
+  if (credentialRecord.credentials.length > 0) {
+    return CredentialRole.Holder
+  }
+  // If credentialRecord.credentials doesn't have any values, and we're also not in state done it means we're the issuer.
+  else if (credentialRecord.state === CredentialState.Done) {
+    return CredentialRole.Issuer
+  }
+  // For these states we know for certain that we're the holder
+  else if (holderCredentialStates.includes(credentialRecord.state)) {
+    return CredentialRole.Holder
+  }
+
+  // Roles only the issuer can have, as sender, when issuing a credential
+  const didCommCredentialIssuerAsSenderRoles = [CredentialState.OfferSent, CredentialState.CredentialIssued]
+
+  // Roles only the holder can have, as sender, when issuing a credential
+  const didCommCredentialHolderAsSenderRoles = [CredentialState.ProposalSent, CredentialState.RequestSent]
+
+  // Roles only the issuer can have, as a recipient, when issuing a credential
+  const didCommCredentialIssuerAsReceiverRoles = [CredentialState.RequestReceived, CredentialState.ProposalReceived]
+
+  // Roles only the holder can have, as a recipient, when issuing a credential
+  const didCommCredentialHolderAsReceiverRoles = [CredentialState.OfferReceived, CredentialState.Done]
+
+  // Fetch the associated didcomm message record
+  const didCommMessageRepository = agent.dependencyManager.resolve(DidCommMessageRepository)
+  const didCommMessageRecord = await didCommMessageRepository.findSingleByQuery(agent.context, {
+    associatedRecordId: credentialRecord.id,
+  })
+
+  if (didCommMessageRecord) {
+    // Check whether the role of the didcomm message is `receiver`
+    if (didCommMessageRecord.role === DidCommMessageRole.Receiver) {
+      // Check whether the state of the credential record is included in the allowed roles for the issuer
+      if (didCommCredentialIssuerAsReceiverRoles.includes(credentialRecord.state)) {
+        return CredentialRole.Issuer
+      }
+
+      // Check whether the state of the credential record is included in the allowed roles for the holder
+      if (didCommCredentialHolderAsReceiverRoles.includes(credentialRecord.state)) {
+        return CredentialRole.Holder
+      }
+    }
+
+    // Check whether the role of the didcomm message is `sender`
+    if (didCommMessageRecord.role === DidCommMessageRole.Sender) {
+      // Check whether the state of the credential record is included in the allowed roles for the issuer
+      if (didCommCredentialIssuerAsSenderRoles.includes(credentialRecord.state)) {
```

Comment:
In all cases except done we can extract the role based on state, so this feels a bit redundant.
We can just check on the offer message, if role == sender, we're role = issuer else role = holder.

Might need to handle the case where there's only a proposal or only a request (if state is abandoned)